QStore4S/src/main/java/edu/asu/qstore4s/controller/QStore.java

```diff
@@ -199,35 +207,79 @@ public String getData(HttpServletRequest request, HttpServletResponse response,
         }
 
-        Class<?> clazz = classMap.get(clas);
-        if (clazz == null) {
-            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
-            return new Error("Please provide a valid Class to execute the request").toString(accept);
+        Map<String, String> responseParams = new HashMap<>();
+        try {
+            int queryID = query.hashCode();
+            responseParams.put("pollurl", "/query/" + queryID);
+            ExecutionStatus queryStatus = asyncQueryManager.getQueryStatus(query.hashCode());
+
+            // if the query is running or completed notify the client
+            if (queryStatus == ExecutionStatus.RUNNING || queryStatus == ExecutionStatus.COMPLETED) {
+                response.setStatus(HttpServletResponse.SC_OK);
+                responseParams.put("queryStatus", queryStatus.name());
+                return new Message(responseParams).toString(accept);
+            }
+
+            Class<?> clazz = classMap.get(clas);
+            if (clazz == null) {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                return new Message("Please provide a valid Class to execute the request").toString(accept);
+            }
+
+            if (accept != null && accept.equals(RETURN_JSON)) {
+                repositorymanager.executeQueryAsync(query, clazz, JSON, queryID);
+            } else {
+                repositorymanager.executeQueryAsync(query, clazz, XML, queryID);
+            }
+
+            response.setStatus(HttpServletResponse.SC_OK);
+            responseParams.put("queryStatus", ExecutionStatus.RUNNING.name());
+            return new Message(responseParams).toString(accept);
+        } catch (ExecutionException e) {
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            responseParams.put("queryStatus", ExecutionStatus.FAILED.name());
+            return new Message(responseParams).toString(accept);
         }
 
-        String returnString = "";
+    }
 
-        if (accept != null && accept.equals(RETURN_JSON)) {
-            returnString = repositorymanager.executeQuery(query, clazz, JSON);
-        } else {
-            returnString = repositorymanager.executeQuery(query, clazz, XML);
+    @ResponseBody
+    @RequestMapping(value = "/query/{queryID}", method = RequestMethod.GET)
+    public String getAysncQueryResult(@RequestHeader("Accept") String accept, @PathVariable("queryID") Integer queryID)
+            throws ExecutionException {
+        ExecutionStatus queryStatus = asyncQueryManager.getQueryStatus(queryID);
+
+        Map<String, String> responseParams = new HashMap<>();
+        responseParams.put("pollurl", "/query/" + queryID);
+        responseParams.put("queryStatus", queryStatus.name());
+
+        if (queryStatus == ExecutionStatus.COMPLETED) {
+            responseParams.put("result", convertToCData(asyncQueryManager.getQueryResult(queryID)));
```

Comment:
Our output xml looks something like this <message><result><?xml?>...
However, The XML Tag should always be on top otherwise we will get this error http://stackoverflow.com/questions/19889132/error-the-processing-instruction-target-matching-xxmmll-is-not-allowed. That's the reason I am parsing it as cdata.